src/components/jscommon/CanvasWidget/CanvasWidgetNew.tsx

```diff
@@ -83,41 +83,46 @@ interface Props<T extends BaseLayerProps> {
 
 const CanvasWidget = <T extends BaseLayerProps>(props: Props<T>) => {
     const divRef = React.useRef<HTMLDivElement>(null)
-    const [state, dispatch] = useReducer(dragReducer, {
-        dragging: false,
-        dragAnchor: undefined,
-        dragPosition: undefined,
-        dragRect: undefined
+    const [dragState, dispatchDrag] = useReducer(dragReducer, {
+        dragging: false
     })
+    const [prevDivElement, setPrevDivElement] = useState<HTMLDivElement | null>(null)
+    const [prevDragState, setPrevDragState] = useState<DragState | null>(null)
+    const [prevWidth, setPrevWidth] = useState<number>(0)
+    const [prevHeight, setPrevHeight] = useState<number>(0)
 
     useEffect(() => {
-        // this is only needed if the previous repaint occurred before the canvas element was rendered to the browser
+        const divElement = divRef.current
         props.layers.forEach((L, index) => {
-            // /// TODO: figure out what to do here
-            // console.log(`I am layer ${index} and my props were ${JSON.stringify(props)}`)
-            const divElement = divRef.current
-            if (divElement) {
+            L.updateProps(props.widgetProps)
+            if ((divElement) && (divElement !== prevDivElement)) {
+                // only repaint if we have a new div element
                 const canvasElement = divElement.children[index]
                 L.resetCanvasElement(canvasElement)
-                L.updateProps(props.widgetProps)
-                L.scheduleRepaint() // The dependency list means this only gets called if props/canvas changed.
-                // In those cases, we definitely need to repaint, so scheduling it should be safe.
+                L.scheduleRepaint()
+            }
+            if ((props.widgetProps.width !== prevWidth) || (props.widgetProps.height !== prevHeight)) {
+                // repaint if dimensions have changed
+                L.scheduleRepaint()
             }
         })
-    }, [props, divRef])
-
-    const _dispatchDragEvents = useCallback((released: boolean) => {
-        if (!state.dragRect) return
-        const pixelDragRect = {
-            xmin: state.dragRect[0],
-            xmax: state.dragRect[0] + state.dragRect[2],
-            ymin: state.dragRect[1] + state.dragRect[3],
-            ymax: state.dragRect[1]
-        }
-        for (let l of props.layers) {
-            l.handleDrag(pixelDragRect, released, state.shift, state.dragAnchor, state.dragPosition)
+        if ((dragState) && (dragState !== prevDragState) && (dragState.dragRect)) {
```

Comment:
Do we need to do a deep compare on this? Or is the state hook making it effectively immutable/you'd wind up replacing the object if any of the properties are changed?